flux-kube/cmd/flux-kube.py

```diff
@@ -49,65 +44,400 @@ def get_parser(self):
+                                               stdin=ps.stdout, 
+                                               stderr=subprocess.STDOUT))
+            ps.wait()
+        except subprocess.CalledProcessError as e:
+            if 'AlreadyExists' in str(e.stdout):
+                pass
+            else:
+                self.logger.error(e)
+                raise
 
 
+class CancelCmd(TranslateCmd):
+    """
+    CancelCmd translates a Flux jobspec V1
+    into the corresponding K8s/OpenShift template 
+    and applies template overrides. Then it 
+    deletes the corresponding objects via `oc delete`.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def main(self, args):
+        super().main(args)
+
+        cmd = ['oc', 'process', args.namespace + '//' + 
+               self.service['template']['metadata']['name']]
+
+        for k, v in self.service['overrides'].items():
+            cmd.extend(('-p', k + '=' + v))
+
+        try:
+            ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+            (subprocess.check_output(('oc', 'delete', '-f', '-'),
+                                      stdin=ps.stdout,
+                                      stderr=subprocess.STDOUT))
+            ps.wait()
+        except subprocess.CalledProcessError as e:
+            if "Error from server (NotFound)" in str(e.output):
+                self.logger.warning("Can't cancel nonexistent objects:\n"
+                                    + str(e.output))
+            else:
+                self.logger.error(e)
+                raise
+        except Exception as e:
+            self.logger.error(e)
+            raise
+
+
+
+class MiniCmd(KubeCmd):
+    """
+    MiniCmd takes a template name and 
+    overrides as arguments and constucts a skeleton 
+    jobspec based on the arguments. It then invokes 
+    SubmitCmd to create the objects or CancelCmd 
+    to delete them.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+        self.logger = logging.getLogger("flux-kube")
+        self.parser.add_argument(
+            "mini",
+            type=str,
+            metavar="M",
+            help="mini command: submit or cancel",
+        )
+        self.parser.add_argument(
+            "template_name",
+            type=str,
+            metavar="T",
+            help="template to submit/cancel",
+        )
+        self.parser.add_argument(
+            "-o",
+            "--overrides",
+            type=str,
+            metavar="OR",
+            default="{}",
+            help="template overrides",
+        )
+
+    def main(self, args):
+        jobspec = '{"version": 1, "resources": '\
+        '[{"type": "slot", "count": 1, "with": '\
+        '[{"type": "core", "count": 1}], "label": "task"}], '\
+        '"tasks": [{"command": ["' + args.template_name + '"], '\
+        '"slot": "task", "count": {"per_slot": 1}}], '\
+        '"attributes": {"system": {"cwd": "/usr/src", "flux-kube": 1, '\
```

Comment:
If the `cwd` of `/usr/src` doesn't exist on rank 0 (where this runs after the alloc bypass), will that cause an issue?

Reply:
No, I set it to a nonexistent directory and it worked fine.

Reply:
It also works just fine without `cwd`, so I'm going to remove that.